atd-vze/src/Components/GridExportData.js

```diff
@@ -0,0 +1,166 @@
+        // Remove __typename from export (contains table name which is already in filename)
+        if (key === "__typename") {
+          delete row["__typename"];
+        } else if (Array.isArray(value)) {
+          // If value is array, recursive call and handle objects in array
+          value = formatExportData(value);
+          value.forEach(object => {
+            Object.entries(object).forEach(([key, value]) => {
+              if (row[key]) {
+                // If top level already has this key, concat
+                row[key] = `${row[key]}, ${value}`;
+              } else {
+                // Else use spread to add to top level
+                row = { ...row, ...object };
+              }
+            });
+            // Delete nested data after added to top level
+            delete row[key];
+          });
+        } else if (typeof value === "object" && value !== null) {
+          // If value is object, remove __typename and move to top level, then delete
+          "__typename" in value && delete value["__typename"];
+          row = { ...row, ...value };
+          delete row[key];
+        }
+      });
+      return row;
+    });
+    return newData;
+  };
+
+  return (
+    <>
+      <StyledSaveLink>
+        <i
+          className="fa fa-save fa-2x ml-2 mt-1"
+          onClick={toggleModalAndExport}
+        />
+      </StyledSaveLink>
+      <Modal isOpen={isModalOpen} toggle={toggleModal} className={"modal-sm "}>
+        <ModalHeader toggle={toggleModal}>
+          Export to .csv ({query.limit} rows)
+        </ModalHeader>
+        <ModalBody>
+          <FormGroup row>
+            <Col sm="8">
+              <Input
+                id="csv-number-input"
+                type="number"
+                placeholder="Number of rows"
+                min={0}
+                max={totalRecords}
+                onChange={setExportLimit}
+              />
+            </Col>
+            <Col sm="4">
+              <Input
+                className="form-check-input"
+                type="checkbox"
+                id="csv-checkbox-input"
+                name="inline-checkbox1"
+                value="option1"
+                onChange={setExportLimit}
+              />
+              <Label
+                className="form-check-label"
+                check
+                htmlFor="inline-checkbox1"
+              >
+                All ({totalRecords})
+              </Label>
+            </Col>
+          </FormGroup>
+        </ModalBody>
+        <ModalFooter>
+          {!loading && data ? (
+            <CSVLink
+              className=""
+              data={formatExportData(data[query.table])}
+              filename={query.table + moment(Date.now()).format()}
+            >
+              <Button color="primary" onClick={toggleModal}>
+                Save
+              </Button>
+            </CSVLink>
+          ) : (
+            <Spinner className="mt-2" color="primary" />
```

Comment:
I'm wondering if it would be at all possible to user a progress bar here instead of a spinner. I guess you would need the `<CSVLink>` component to be signaling how far along it is in processing the data which might not be possible, but just a thought....

https://localhost:3000/#/dev/base/progress-bar

Reply:
I guess we could also fake it by creating an estimate based on the number of records and filling up the progress bar based on that estimate... Not sure how most people do this tbh.

Reply:
Yeah! I'll look into this, create an issue, and estimate the effort it would take. That would be a really cool improvement over staring at the spinner.